shapeshift/web
# feat(cosmos): claim staking and unstaking

## Description

This implements the building of unsigned claim Tx for cosmos and broadcasting of the Tx with native wallet.

To be added in a follow-up PR(s):

- Broadcast errors (will require working WS txhistory messages)
- Don't claim 0 rewards, since it is possible at protocol level to broadcast such Tx but no rewards will effectively be claimed (it could be abstracted to something similar to `useSendDetails`) - we will probably need to add some rounding logic in a follow-up PR, since we never really get `0` as rewards amount, but rather a very small number
- Don't allow undelegations if greater than `MaxEntries` (https://docs.cosmos.network/master/modules/staking/03_messages.html) - we would need to expose this in the unchained response for `/api/v1/account/{pubkey}`, and then add it to the parsed data in chain-adapters before we can consume it in web. When that's done, it could be abstracted to something similar to `useSendDetails`.

## Notice




## Pull Request Type


## Issue (if applicable)

- closes #657
- closes #664
- closes #666
- closes #667
- closes #669 
- closes #670
- closes #671

## Risk

N/A, this is a separate domain. Errors are not properly handled yet and will need to as a follow-up.

## Testing

- Go to Cosmos Staking overview on an account with some rewards
- Click claim and continue with the claim flow
- It should display loading state while Tx is built/broadcasted 
- It should broadcast Tx and change the broadcast button to done/display the Txid

## Screenshots (if applicable)

### Claim
![image](https://user-images.githubusercontent.com/17035424/161162300-10e7863b-4af4-4152-823f-04035788df0f.png)
![image](https://user-images.githubusercontent.com/17035424/161162342-05ed0a78-4721-4aa7-9b82-f199114afa69.png)
![image](https://user-images.githubusercontent.com/17035424/161162374-877af1da-0713-4d81-b756-8667557cb4e9.png)
![image](https://user-images.githubusercontent.com/17035424/161162385-21926924-ce37-46b5-8dba-4f3c54d4f30d.png)

### Stake

<img width="477" alt="image" src="https://user-images.githubusercontent.com/17035424/161319015-0b22b176-c8f4-4731-bde0-4a9c1577a38d.png">
<img width="493" alt="image" src="https://user-images.githubusercontent.com/17035424/161319058-64c3fbc8-1a08-4f10-b205-8c120f53a997.png">
<img width="461" alt="image" src="https://user-images.githubusercontent.com/17035424/161319131-b806412e-78a4-4c47-b10b-1992a5402bc7.png">
<img width="500" alt="image" src="https://user-images.githubusercontent.com/17035424/161319191-30e2032d-19f6-4c06-877e-0fc2ce5850aa.png">
<img width="458" alt="image" src="https://user-images.githubusercontent.com/17035424/161319222-9060c02c-b7f1-4063-8ec8-37dd216d704e.png">

### Unstake

<img width="454" alt="image" src="https://user-images.githubusercontent.com/17035424/161319300-e367a1b9-e65a-41ff-b080-4cbdecb47c13.png">
<img width="464" alt="image" src="https://user-images.githubusercontent.com/17035424/161319584-45cfbf0e-23d3-47f2-ada2-18936835db1c.png">
<img width="458" alt="image" src="https://user-images.githubusercontent.com/17035424/161320836-dcbd1f01-b57f-4ce4-a2a4-196304967e4a.png">
<img width="484" alt="image" src="https://user-images.githubusercontent.com/17035424/161323188-2ceeec15-a8f0-4a98-b066-e025c73fd2dc.png">
<img width="458" alt="image" src="https://user-images.githubusercontent.com/17035424/161320898-e91a34af-42c9-4fe7-9682-84609d3e6105.png">
